Getting started guide (#15)
neo4j/cypher-builder: docs/modules/ROOT/pages/getting-started/filters-and-projections.adoc

= Filters and projections

> After making a basic `MATCH` & `RETURN` Cypher query in xref:getting-started/your-first-query.adoc[Your First Query], in this guide you'll add some filters and change the projection to only return some fields.

Reviewer: Sorry, for some reason I skipped this file during my review, so I'm catching up now.

```suggestion
This tutorial shows how to add filters and change the projection to return specific fields. 
Refer to xref:getting-started/your-first-query.adoc[Your First Query] for more information on how to write queries.
```

^ I made a suggestion about the name of the page Your First Query, but I kept it the way it is right now so we can check later.

Reviewer: It would be good to refer here that params are also being addressed in this page. I'm not sure how to suggest that though.

The resulting query will look like this:

```cypher
MATCH(m:Movie)
WHERE (m.title = "The Matrix" AND m.released < 2000)
RETURN m.title, m.tagline, m.released
```

NOTE: This query will return all movies with the name `"The Matrix"` released before `2000`. This query is based on the default example database in Neo4j, and will return the classic 1999 film **The Matrix**. 

== Filtering

When creating a `MATCH` clause, it is possible to add subclauses to it in a similar way to how you added the `RETURN` statement. Change the `Cypher.Match` instance to add a `WHERE` subclause with the method `.where`:

```javascript
const clause = new Cypher.Match(movieNode)
    .where(Cypher.eq(movieNode.property("title"), new Cypher.Param("The Matrix")))

>     .return(movieNode);

Reviewer 2: Open question: Maybe we can structure the filtering by introducing the concept of predicate, in that way, we can act in isolation by:
- how to use `Cypher.eq` to build the predicate `titleEqMatrix`
- how to use `Cypher.lt` to build the predicate `releasedBefore2000`
- how to use `Cypher.and`

At this point, we can say something "The last part is to using this predicate by append `.where` ..."

Reviewer: I think we might be missing some explanation about `new` when used in the context of `new Param`

Author: @MacondoExpress I've tried to avoid adding explanation and internal details on this Getting Started guide, once a page for predicates with more detail is added we could link to it from here

```

> There is quite a bit to unwrap here. First, you'll notice that the `.where` method takes the result of the function `Cypher.eq`. This function is the **eq**uality (`=`) operator that compares the property `title` and the string `The Matrix`.

Reviewer: @MacondoExpress 's suggestion above could be relevant here so the unwrapping is not that hard or even necessary

`Cypher.eq` takes two arguments to compare. Our first argument will be the property `title`. Because our title is a property of our movieNode variable, the `.property` method of that node will ensure that the property correctly references the movieNode variable.

The second parameter is `"The Matrix"`, but instead of using a plain string, the object `Cypher.Param` class will take care of turning that into a parameter so we can safely put user input here.

> .Why `Cypher.eq` doesn't need `new`?

Reviewer: As a matter of style guide, we don't use titles for notes

```suggestion
```

[NOTE]
====

> Cypher Builder treats Variables (like nodes and relationships) and Clauses as classes, as these are intended to be assigned to a JavaScript variable to be reused, composed and built.

Reviewer: ```suggestion
Similarly to nodes and relationships, Cypher Builder treats variables and clauses as classes.
This is because they are intended to be reused, composed, and built.
```

Cypher Functions and Procedures, however, are intended to be used a single time, as part of a larger query, so Cypher Builder treats these as simple functions, making it closer to how these work in Cypher.
====

=== Cypher and params

> You can run your script again (`node main.js`), and should get the following Cypher:

Reviewer: ```suggestion
By running the script again (`node main.js`), you should get the following Cypher:
```

```cypher
MATCH (this0:Movie)
WHERE this0.title = $param0
RETURN this0
```

> Success! The movie is now filtered by its title to a param. The param, however, is missing from the output. You'll need to tweak the build process at the end of the script to get the parameters as well:

Reviewer: ```suggestion
That means `Movie` is now filtered by its title to a param, but the param itself is missing from the output.
To fix that, you need to modify the build process at the end of the script:
```

```javascript
const { cypher, params } = match.build();
console.log(cypher);
console.log(params);
```

> Now, executing again, we get all we need to run this query in Neo4j:

Reviewer: ```suggestion
After executing it, you can run the following query in Neo4j:
```

```cypher
MATCH (this0:Movie)
WHERE this0.title = $param0
RETURN this0
```

```javascript
{ param0: 'The Matrix' }
```

[TIP]
====

> Note that `new Param` already takes care of assigning a param name and correctly matching the value to that parameter in the Cypher query. If a `Param` instance is used multiple times within a query, it will always reference the same parameter (`$param0` in this case).

Reviewer: The absence of explanation about `new Param` before is making this note a bit difficult to understand

====

=== `AND` filtering

> The query still lacks the second filter, which needs to be concatenated with `AND`. Simply use the method `.and` after the `where` with a new `Cypher.eq` predicate:

Reviewer: ```suggestion
To add a second filter to the query, for instance `AND`, use the method `.and` after `.where` with a new `Cypher.eq` predicate:
```

```javascript

>     .where(Cypher.eq(movieNode.property("title"), new Cypher.Param("The Matrix")))
>     .and(Cypher.lt(movieNode.property("released"), new Cypher.Param(2000)))

Reviewer 2: I believe this could be opinionated, maybe @lidiazuin  knows better here, but I believe that providing these steps at this level of isolation increases the cognitive complexity for the reader, also a more experienced dev who comes from others SQL/Cypher builders will use the getting started as a fast overview of the product without actually have an incremental script file in front. What I'm trying to say is that maybe it's better to not break the method chain in the step, for instance:

```suggestion
const clause = new Cypher.Match(movieNode)
  .where(Cypher.eq(movieNode.property("title"), new Cypher.Param("The Matrix")))
  .and(Cypher.lt(movieNode.property("released"), new Cypher.Param(2000)))
  .return(movieNode);
```

Reviewer: Well, the getting started part is mostly for people who are just getting to start using a feature, so it's good to have more explanation and contextualization. I just don't know if that is going too much into detail or not, if we indeed need to break up a little bit to make things more understandable to more inexperienced readers.

Author: Here I agree with @lidiazuin
At least my intention here was to bring someone with little knowledge to be able to use the library and it's main features. Other guides like how-tos should definitely be more concise IMO

Reviewer: If we stick to the definition of guides and tutorials proposed by Diataxis, we can understand that tutorials are designed for new users who need to learn what's all about and a guide is for a more experienced reader. A comparison would be a tutorial on how to start cooking and a guide is a recipe for chocolate cake 😊

Reviewer 2: Sure, anyway it's already a lot of effort writing these kinds of documentation, so read my comment as a genuine curiosity about the style and not as a request for change!

Reviewer 2: The sentiment behind the comment was that this guide is structured in the form of:
- incremental step
- incremental step
- ...incremental step
- conclusion.

**For a very beginner dev**: The only positive feedback happens at the conclusion step. If the user made a mistake, he needs to reach the conclusion step to recognize it.

**For a very expert user**: He already started with the expectation to find something along `Match(entity).where(predicate)` or `Select(columns).from(entity).build()`, so potentially (**bad** 🙊 ) immediately jump to the conclusion section, with the risk to missing some interesting catch.

But is a comment completely driven by sentiment, so happy if I'm wrong!

```

This takes the same parameters as `.where`, and makes sure to properly concatenate the filters with `AND`. In this case, the function `Cypher.lt` will add the operator `<` (**l**ower **t**han). Same as before, the comparison happens between a property of `movieNode` and a param. This time the param is a number. Cypher Builder will take care of correctly passing it to the `params` object.

Running the script again:

```cypher
MATCH (this0:Movie)
WHERE (this0.title = $param0 AND this0.released < $param1)
RETURN this0
```

```javascript
{ param0: 'The Matrix', param1: 1999 }
```

This should look closer to the target query, note that now there are two parameters, each one correctly associated to the correct filter.

[NOTE]
====
Adding the `AND` operation automatically adds parenthesis to ensure operation precedence, this will be important when dealing with complex and nested filters 
====

> == Projection

Reviewer: I have the feeling this section should be mentioned in the introductory paragraph and maybe have some explanation of what projection means or is this a common term?

Author: It is briefly mentioned: `in this guide you'll add some filters and change the projection to only return some fields`

I assume people are roughly familiar with what projection means

> Usually, returning the full node is not desirable, for these cases, you can explicitly set a projection in the `RETURN` clause.

Reviewer: Could "desirable" be replaced by "necessary" here?

```suggestion
Usually, returning the full node is not necessary.
In cases like this, you can explicitly set a projection in the `RETURN`.
```

> You can define multiple columns to return in the projection by passing them into the `.return` method. This method not only accepts Nodes, but any kind of variables, such as properties or params.

Reviewer: ```suggestion
For that, you can define multiple columns to return in the projection by passing them into the `.return` method.

[NOTE]
====
This method accepts any kind of variables, such as properties or params.
===
```

> Let's update the return subclause:

Reviewer: ```suggestion

```

```javascript
    .return(movieNode.property("title"), movieNode.property("tagline"), movieNode.property("year"));
```

> The returned Cypher now should be:

Reviewer: ```suggestion
The returned Cypher now should feature a query with filters and projections:
```

```cypher
MATCH (this0:Movie)
WHERE (this0.title = $param0 AND this0.released < $param1)
RETURN this0.title, this0.tagline, this0.released
```

> Success! You have now a query with filters and projections.

Reviewer: ```suggestion
```

=== Reusing variables

> Right now, this query is filtering and returning the properties `id` and `title` of the Movie nodes. However, the query relies on the name of these properties to remain the same. This, particularly with more complex queries, can lead to flaky queries, as it is hard to keep track of strings identifying values.

Reviewer: ```suggestion
The current query is filtering and returning the properties `id` and `title` of the `Movie` nodes. 
However, the query relies on the name of these properties to remain the same. 
In more complex complex, this strategy could lead to flaky queries, as it is hard to keep track of strings identifying values.
```

> To fix this, you'll make a small adjustment to the script, so that both the filters and the projection use the same property variable:

Reviewer: ```suggestion
To fix that, you need to ensure both filters and projection use the same property variable:
```

```javascript
const titleProp = movieNode.property("title");
const yearProp = movieNode.property("released");
const taglineProp = movieNode.property("tagline");

const clause = new Cypher.Match(movieNode)
    .where(Cypher.eq(titleProp, new Cypher.Param("The Matrix")))
    .and(Cypher.lt(yearProp, new Cypher.Param(2000)))
    .return(titleProp, taglineProp, yearProp);
```

> This change, not only ensures that different parts of the query are always in sync but also makes the clause itself shorter and easier to read.

Reviewer: In general, we try to avoid the word "easy" or judgements like that as it could be not that easy for some users

```suggestion
That should keep different parts of the query in sync and also make the clause itself shorter and quicker to read.
```

[TIP]
====

> Params can also be assigned to a variable and reused, this can be particularly useful when having multiple filters over the same parameter.

Reviewer: This leads me to a question whether we should use parameter or param when referring to this topic here.

```suggestion
Params can also be assigned to a variable and be reused. 
This can be particularly useful when having multiple filters over the same parameter.
```

====

== Conclusion

> Your script should now look similar to this:

Reviewer: ```suggestion
After going through all the steps previously described, your script should now look similar to this:
```

```javascript
import Cypher from "@neo4j/cypher-builder";

const movieNode = new Cypher.Node({
    labels: ["Movie"],
});

const titleProp = movieNode.property("title");
const yearProp = movieNode.property("released");
const taglineProp = movieNode.property("tagline");

const clause = new Cypher.Match(movieNode)
    .where(Cypher.eq(titleProp, new Cypher.Param("The Matrix")))
    .and(Cypher.lt(yearProp, new Cypher.Param(2000)))
    .return(titleProp, taglineProp, yearProp);

const { cypher, params } = clause.build();
console.log(cypher);
console.log(params);
```

And its execution should show the following query:

```cypher
MATCH (this0:Movie)
WHERE (this0.title = $param0 AND this0.released < $param1)
RETURN this0.title, this0.tagline, this0.released
```

```javascript
{ param0: 'The Matrix', param1: 2000 }
```

> With this, you already have the tools to make useful (albeit simple) queries and have also learnt how to deal with parameters. Feel free to try other filters operations (TODO: add link )

Reviewer 2: ```suggestion
With this, you already have the tools to make useful (albeit simple) queries and have also learnt how to deal with parameters.
```

Reviewer: I would add this information at the top of the page, not here, so readers know from the beginning what they will be achieving with this tutorial

> In the next tutorial, you'll add relationships and more complex filtering to this query.

Reviewer: Same, I would add this in the beginning and then link to the other page about more complex filtering.
